### PR TITLE
[CRDB-9007] sql: allow users to view SHOW RANGES with ZONECONFIG privilege

### LINKED ISSUE #68369: The ability to GRANT "SHOW RANGES" for a role
**Is your feature request related to a problem? Please describe.**
We can provide roles with the permission ZONECONFIG, however to debug ranges it is required to be an admin.  As our customer does not want to make all users admin, our customers require the ability to  add `SHOW RANGES` to a role.  

**Describe the solution you'd like**
Create a new role permission that allows users to `SHOW RANGES`

**Describe alternatives you've considered**
Giving admin to users, which is not possible in this customers environment. 

**Additional context**

For example, I created a role where we can alter zone config, but can't debug/verify the ranges using `SHOW RANGES`

`operator@:26257/defaultdb> ALTER TABLE defaultdb.test CONFIGURE ZONE USING lease_preferences = '[[+region=local]]',constraints = '{"+region=local": 3, "+region=node3region": 2}', num_replicas=5;
CONFIGURE ZONE 1
operator@:26257/defaultdb> SHOW RANGES from database defaultdb;
ERROR: only users with the admin role are allowed to read crdb_internal.ranges_no_leases
`

gz#9363

### PR DESCRIPTION
**Pull request by Santamaura**
fixes https://github.com/cockroachdb/cockroach/issues/68369

Previously the SHOW RANGES query was limited to users with the admin role.
This was an issue because customers did not want to allow many users access
to the admin role. Therefore, the code has been altered so that users can now either
have access to the admin role or have the ZONECONFIG privilege in order to
view the ranges.

Release note (sql change): allow users who are not admin to SHOW RANGES if ZONECONFIG is granted

User (operator) who is not a member of admin role and doesn't have `ZONECONFIG` privilege:
Screen Shot 2022-01-26 at 5 53 19 PM
Screen Shot 2022-01-26 at 5 55 03 PM

After granting `ZONECONFIG`:
Screen Shot 2022-01-26 at 5 55 18 PM
Screen Shot 2022-01-26 at 5 55 40 PM

### CODE REVIEW

File: pkg/sql/crdb_internal.go [hunk 1]
```diff
@@ -3028,9 +3028,6 @@ CREATE TABLE crdb_internal.ranges_no_leases (
 )
 `,
 	generator: func(ctx context.Context, p *planner, _ catalog.DatabaseDescriptor, _ *stop.Stopper) (virtualTableGenerator, cleanupFunc, error) {
-		if err := p.RequireAdminRole(ctx, "read crdb_internal.ranges_no_leases"); err != nil {
```
Comment (RichardJCai) [inline pkg/sql/crdb_internal.go:3031]: We should make it such that if they have admin OR `ZONECONFIG` they can perform this.

root does not have `ZONECONFIG` on system tables it seems.

```
demo@127.0.0.1:26257/movr> show grants on system.jobs for root;
  database_name | schema_name | table_name | grantee | privilege_type
----------------+-------------+------------+---------+-----------------
  system        | public      | jobs       | root    | DELETE
  system        | public      | jobs       | root    | GRANT
  system        | public      | jobs       | root    | INSERT
  system        | public      | jobs       | root    | SELECT
  system        | public      | jobs       | root    | UPDATE
```

Comment (Santamaura) [reply pkg/sql/crdb_internal.go:3031]: Thanks for the early review Richard! Yes either having the admin role or `ZONECONFIG` makes more sense, I've made the changes.

**Comment by Santamaura** [comment]
Just posting my thoughts as of now: I'm not sure if it's possible to permission regulate the use of `crdbInternalRangesNoLeasesTable` with the granularity of a privilege like `ZONECONFIG` as privileges are checked on a per descriptor basis. For e.g. if there exists table `A` and table `B` and a new role is created that is not admin and has `ZONECONFIG` privilege on `A`, it is not immediately apparent how within the context of this function we can filter out `B` if the user wants to view ranges on `B`  (through methods other than the sql shell as a new check has been added for this case)

File: pkg/sql/crdb_internal.go (hunk 1 above)
Comment (Santamaura) [reply pkg/sql/crdb_internal.go:3031]: @RichardJCai I'm starting to think it may not be possible, see my comment below

**Comment by Santamaura** [comment]
adding Marylia per Nam's suggestion in case she has familiarity/thoughts on the changes

**Comment by maryliag** [comment]
I recently created a new role and the idea was similar to this, so if they had any of two specific roles they had access to things, so I created a new require check for that case that you can take a look [here](https://github.com/cockroachdb/cockroach/pull/74715/files#diff-1402bca7aa2ae9e5b623ab3d1e93d918e0f8d652c689c42d644e46611151c9f6R3110-R3137). But that is mostly focus on giving/restricting access on things on DB Console. I don't have enough knowledge on `SHOW RANGES` and `ZONECONFIG` to give an input on how they work (or should work in this case)

**Review by koorosh — COMMENTED** [review]
@santamaura , another challenge that probably has to be taken into consideration is validating that users with granted permission on parent object (ie `database`) also have permissions to child (`schema`, `table`, `index`) objects as well.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75551#-:-MujewfV7Nn_rpNxgNtb:b-ynndk5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @maryliag)

**Review by Santamaura — COMMENTED** [review]
Good call Andrii! I was manually testing this out and it seems the logic implemented should handle these cases as well. Continuing on the example above, if the user has no privilege on the table that they wish to `SHOW RANGES FROM INDEX ...`:

 
and subsequently, when user tries to `SHOW RANGES FROM DATABASE ...`:


There isn't a clear cut way of outputting an error in the database case but it will filter out the rows that the user has no privilege for and I've also added a warning log per our brief discussion this morning.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75551#-:-MulONH17sMrJk0qj_Sm:b-idmbgw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @maryliag)

**Review by rafiss — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75551)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @maryliag and @Santamaura)

___
*[pkg/sql/crdb_internal.go, line 3065 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75551#-Muqav0z34aafaZlh8S6:-Muqav0z34aafaZlh8S7:bujfogg) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4c69427fa4460c9445e3d066de468c67563ce5e/pkg/sql/crdb_internal.go#L3065)):*
> ```Go
> 		for _, desc := range descs {
> 			id := uint32(desc.GetID())
> 			if p.CheckPrivilege(ctx, desc, privilege.ZONECONFIG) != nil && !hasAdmin {
> ```

i'd change this so `hasAdmin` is checked first, so you can avoid the privilege check if it's not needed.

```
if !hasAdmin {
  if err := p.CheckPrivilege(ctx, desc, privilege.ZONECONFIG); err != nil {
    continue
  }
}
```
___
*[pkg/sql/crdb_internal.go, line 3066 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75551#-MuqannO-mecD-CoxVeW:-MuqannO-mecD-CoxVeX:b-zfnvhv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4c69427fa4460c9445e3d066de468c67563ce5e/pkg/sql/crdb_internal.go#L3066)):*
> ```Go
> 			id := uint32(desc.GetID())
> 			if p.CheckPrivilege(ctx, desc, privilege.ZONECONFIG) != nil && !hasAdmin {
> 				log.Warningf(ctx, "user has no ZONECONFIG privilege or admin role for %s", desc.GetName())
> ```

i don't think we need to log this
___
*[pkg/sql/crdb_internal.go, line 3087 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75551#-MuqboN3-n5BPVS8Gkvz:-MuqboN3-n5BPVS8Gkw-:b-o4f0oh) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4c69427fa4460c9445e3d066de468c67563ce5e/pkg/sql/crdb_internal.go#L3087)):*
> ```Go
> 		// if the user has no ZONECONFIG privilege on any table/schema/database
> 		if !hasPermission {
> 			return nil, nil, pgerror.Newf(pgcode.InsufficientPrivilege, "only users with the ZONECONFIG privilege or have the admin role can read crdb_internal.ranges")
> ```

nit: it should say "can read crdb_internal.ranges_no_leases"

also, change to "only users with the ZONECONFIG privilege or the admin role can read..."
___
*[pkg/sql/delegate/show_ranges.go, line 36 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75551#-Muq_Lqt5a71bPM8yJu6:-Muq_Lqt5a71bPM8yJu7:b24fb3m) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4c69427fa4460c9445e3d066de468c67563ce5e/pkg/sql/delegate/show_ranges.go#L36)):*
> ```Go
> 		return err
> 	}
> 	// User needs to either have admin access or have the correct ZONECONFIG privilege
> ```

nit: i'd change this to short-circuit for admins so it's a bit more clear.

and also use `err :=` so that you can wrap the error -- we don't want to lose any of the error cause chain

```
if hasAdmin {
  return nil
}
if err := d.catalog.CheckPrivilege(d.ctx, table, privilege.ZONECONFIG); err != nil {
  return pgerror.Wrapf(err, pgcode.InsufficientPrivilege, "only users with the ZONECONFIG privilege or have the admin role can use SHOW RANGES on %s", table.Name())
}
```

you don't need to use `string()` on the table name
___
*[pkg/sql/delegate/show_ranges.go, line 38 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75551#-Muqc7vB9JEfVpu-6Kxc:-Muqc7vB9JEfVpu-6Kxd:b-cnwxen) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4c69427fa4460c9445e3d066de468c67563ce5e/pkg/sql/delegate/show_ranges.go#L38)):*
> ```Go
> 	// User needs to either have admin access or have the correct ZONECONFIG privilege
> 	if d.catalog.CheckPrivilege(d.ctx, table, privilege.ZONECONFIG) != nil && !hasAdmin {
> 		return pgerror.Newf(pgcode.InsufficientPrivilege, "only users with the ZONECONFIG privilege or have the admin role can use SHOW RANGES on %s", string(table.Name()))
> ```

nit: change to "only users with the ZONECONFIG privilege or the admin role can use SHOW RANGES on %s"
___
*[pkg/sql/logictest/testdata/logic_test/ranges, line 498 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75551#-MuqaRS_Ct2xreICVr-L:-MuqaRS_Ct2xreICVr-M:bdghis5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4c69427fa4460c9445e3d066de468c67563ce5e/pkg/sql/logictest/testdata/logic_test/ranges#L498)):*
> ```
> 
> statement ok
> SHOW RANGES FROM INDEX t@idx
> ```

nit: add an ending newline

**Review by Santamaura — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75551)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @maryliag and @rafiss)

___
*[pkg/sql/crdb_internal.go, line 3065 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75551#-Muqav0z34aafaZlh8S6:-MuquVAE3OnqMnv2HkOv:bd371pr) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4c69427fa4460c9445e3d066de468c67563ce5e/pkg/sql/crdb_internal.go#L3065)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i'd change this so `hasAdmin` is checked first, so you can avoid the privilege check if it's not needed.

```
if !hasAdmin {
  if err := p.CheckPrivilege(ctx, desc, privilege.ZONECONFIG); err != nil {
    continue
  }
}
```
</blockquote></details>

Makes sense to me. Done.
___
*[pkg/sql/crdb_internal.go, line 3066 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75551#-MuqannO-mecD-CoxVeW:-MuquZ6q5wBdZUGAuxtI:bkh3s7t) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4c69427fa4460c9445e3d066de468c67563ce5e/pkg/sql/crdb_internal.go#L3066)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i don't think we need to log this
</blockquote></details>

Fair enough. Wasn't sure if log was necessary here or not.
___
*[pkg/sql/crdb_internal.go, line 3087 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75551#-MuqboN3-n5BPVS8Gkvz:-MuqvEQT-gL3worHkONM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4c69427fa4460c9445e3d066de468c67563ce5e/pkg/sql/crdb_internal.go#L3087)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: it should say "can read crdb_internal.ranges_no_leases"

also, change to "only users with the ZONECONFIG privilege or the admin role can read..."
</blockquote></details>

Done.
___
*[pkg/sql/delegate/show_ranges.go, line 36 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75551#-Muq_Lqt5a71bPM8yJu6:-Muqv0LxBQBWnbjTBYgq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4c69427fa4460c9445e3d066de468c67563ce5e/pkg/sql/delegate/show_ranges.go#L36)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: i'd change this to short-circuit for admins so it's a bit more clear.

and also use `err :=` so that you can wrap the error -- we don't want to lose any of the error cause chain

```
if hasAdmin {
  return nil
}
if err := d.catalog.CheckPrivilege(d.ctx, table, privilege.ZONECONFIG); err != nil {
  return pgerror.Wrapf(err, pgcode.InsufficientPrivilege, "only users with the ZONECONFIG privilege or have the admin role can use SHOW RANGES on %s", table.Name())
}
```

you don't need to use `string()` on the table name
</blockquote></details>

Done.
___
*[pkg/sql/delegate/show_ranges.go, line 38 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75551#-Muqc7vB9JEfVpu-6Kxc:-Muqv8bF3oN8VVfdmqCZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4c69427fa4460c9445e3d066de468c67563ce5e/pkg/sql/delegate/show_ranges.go#L38)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: change to "only users with the ZONECONFIG privilege or the admin role can use SHOW RANGES on %s"
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/ranges, line 498 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75551#-MuqaRS_Ct2xreICVr-L:-MuqvNew7PkJooirAOxz:bqmjwcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4c69427fa4460c9445e3d066de468c67563ce5e/pkg/sql/logictest/testdata/logic_test/ranges#L498)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: add an ending newline
</blockquote></details>

Added.

**Review by rafiss — CHANGES_REQUESTED** [review]
looking close!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75551#-:-MuvdVYV7A7xBEdOQu3R:b1ns4zv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @koorosh, @maryliag, @rafiss, and @Santamaura)

___
*[pkg/sql/delegate/show_ranges.go, line 41 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75551#-MuvdOk58IgI8PbuofoE:-MuvdOk58IgI8PbuofoF:b-bx1hms) ([raw file](https://github.com/cockroachdb/cockroach/blob/05e304939ea5bb4242689a59e5fb29e4d9d71c20/pkg/sql/delegate/show_ranges.go#L41)):*
> ```Go
> 	}
> 	if err := d.catalog.CheckPrivilege(d.ctx, table, privilege.ZONECONFIG); err != nil {
> 		return pgerror.Newf(pgcode.InsufficientPrivilege, "only users with the ZONECONFIG privilege or the admin role can use SHOW RANGES on %s", table.Name())
> ```

nit: change this to `pgerror.Wrapf(err, ...)`

**Review by Santamaura — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75551)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @koorosh, @maryliag, and @rafiss)

___
*[pkg/sql/delegate/show_ranges.go, line 41 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75551#-MuvdOk58IgI8PbuofoE:-MuviL4nCbgwnzcKG0DB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/05e304939ea5bb4242689a59e5fb29e4d9d71c20/pkg/sql/delegate/show_ranges.go#L41)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: change this to `pgerror.Wrapf(err, ...)`
</blockquote></details>

Done.

**Review by rafiss — APPROVED** [review]
this lgtm!

> Reviewed 1 of 3 files at r2, 4 of 5 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75551#-:-MuwPmvl9yr9XXH6sFYX:becqc6o)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @koorosh, @maryliag, and @rafiss)

**Review by koorosh — COMMENTED** [review]
@santamaura , Db Console has `ranges` API that returns ranges info per node (https://github.com/cockroachdb/cockroach/blob/master/pkg/server/serverpb/status.proto#L1482). Does it rely on this change or has to be extended additionally?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75551#-:-MuymMu0ET_qwdBBT1aM:btk1au3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @maryliag and @rafiss)

**Review by Santamaura — COMMENTED** [review]
I was looking through how the `ranges` API works and don't think it will need to be changed/extended. There may be the case where a user who visits that page is both non admin and with no `ZONECONFIG` privilege will instead now get a no value response instead of an error so perhaps in a separate issue we may check if this is the case and handle accordingly.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75551#-:-Mv-TbDE20ErL3JZnN1n:bo603tu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @maryliag and @rafiss)

**Comment by rafiss** [comment]
Hm, to avoid that, is there a way to update this PR so that if it ends up returning zero rows, it creates the error message at that point? (not sure how easy this is with a virtual table generator, just brainstorming)

**Review by Santamaura — COMMENTED** [review]
I was wrong in my above assumption:

Screen Shot 2022-02-03 at 12 43 15 PM

Seems like there is another place in the code that does some admin privilege check anyways so it should be ok!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75551#-:-Mv-dc4f2FPlC0V9Zhta:b-ozpt8a)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @maryliag and @rafiss)

**Comment by Santamaura** [comment]
I'm gonna merge this towards EoD so if there are any last minute comments, comment now 😄

**Comment by Santamaura** [comment]
bors retry

**Comment by Santamaura** [comment]
blathers backport 21.2 21.1

**Pull request merged**